File: src/mobile_robot_localization/src/kalman_filter.cpp
#include "mobile_robot_localization/kalman_filter.hpp"

KalmanFilter::KalmanFilter()
: Node("kalman_filter")
, angular_velocity_z_mean(0.0)
, angular_velocity_z_variance(10.0)
, imu_angular_velocity_z_mean(0.0)
, imu_angular_velocity_z_variance(0.5)
, is_first_odom(true)
, odom_angular_velocity_z_last(0.0)
, angular_velocity_z_motion_mean(0.0)
, angular_velocity_z_motion_variance(4.0)
{
    odom_sub = create_subscription<Odometry>(
        "odom",
        10,
        [this](const Odometry& odom)-> void {odom_callback(odom);}
    );

    imu_sub = create_subscription<Imu>(
        "imu/data_raw",
        10,
        [this](const Imu& imu)-> void {imu_callback(imu);}
    );

    odom_pub = create_publisher<Odometry>(
        "odom_kalman",
        10
    );

    RCLCPP_INFO_STREAM(get_logger(), std::string(get_name()) << " node initialized!");
}

void KalmanFilter::state_prediction()
{
    angular_velocity_z_mean = angular_velocity_z_mean + angular_velocity_z_motion_mean;
    angular_velocity_z_variance = angular_velocity_z_variance + angular_velocity_z_motion_variance;
}

void KalmanFilter::measurement_update()
{
    angular_velocity_z_mean = (imu_angular_velocity_z_variance * angular_velocity_z_mean + angular_velocity_z_variance * imu_angular_velocity_z_mean) / (angular_velocity_z_variance + imu_angular_velocity_z_variance);
    angular_velocity_z_variance = (angular_velocity_z_variance * imu_angular_velocity_z_variance) / (angular_velocity_z_variance + imu_angular_velocity_z_variance);
}

void KalmanFilter::odom_callback(const Odometry& odom)
{
    kalman_odom = odom;

    if(is_first_odom)
    {
        odom_angular_velocity_z_last = odom.twist.twist.angular.z;
        angular_velocity_z_mean = odom.twist.twist.angular.z;
        is_first_odom = false;
        return;
    }

    angular_velocity_z_motion_mean = odom.twist.twist.angular.z - odom_angular_velocity_z_last;
    if(!std::isfinite(angular_velocity_z_motion_mean))
    {
        return;
    }

    state_prediction();

    measurement_update();

    kalman_odom.twist.twist.angular.z = angular_velocity_z_mean;
    odom_pub->publish(kalman_odom);

    // RCLCPP_INFO_STREAM(get_logger(), "=============================");
    // RCLCPP_INFO_STREAM(get_logger(), "angular z mean: " << angular_velocity_z_mean);
    // RCLCPP_INFO_STREAM(get_logger(), "imu angular z mean: " << imu_angular_velocity_z_mean);
    // RCLCPP_INFO_STREAM(get_logger(), "odom angular z: " << odom.twist.twist.angular.z); // This comes from w in noisy_odom.cpp and sometimes is nan
    // RCLCPP_INFO_STREAM(get_logger(), "motion angular z mean: " << angular_velocity_z_motion_mean);
    // RCLCPP_INFO_STREAM(get_logger(), "angular z variance: " << angular_velocity_z_variance);
}

void KalmanFilter::imu_callback(const Imu& imu)
{
    imu_angular_velocity_z_mean = imu.angular_velocity.z;
}


int main(int argc, const char * const * argv)
{   
    rclcpp::init(argc, argv);
    auto node = std::make_shared<KalmanFilter>();
    rclcpp::spin(node);
    node.reset();
    rclcpp::shutdown();

    return 0;
}

File: src/mobile_robot_localization/mobile_robot_localization/kalman_filter.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Imu
import math

class KalmanFilter(Node):
    def __init__(self):
        super().__init__(node_name="kalman_filter")

        self._odom_sub = self.create_subscription(
            msg_type=Odometry,
            topic="odom",
            callback=self._odom_callback,
            qos_profile=10,
        )
        self._imu_sub = self.create_subscription(
            msg_type=Imu,
            topic="imu/data_raw",
            callback=self._imu_callback,
            qos_profile=10,
        )
        self._odom_pub = self.create_publisher(
            msg_type=Odometry,
            topic="odom_kalman",
            qos_profile=10,
        )

        self._angular_velocity_z_mean = 0.0
        self._angular_velocity_z_variance = 10.0

        self._imu_angular_velocity_z_mean = 0.0
        self._imu_angular_velocity_z_variance = 0.5

        self._is_first_ododm = True

        self._odom_angular_velocity_z_last = 0.0
        self._angular_velocity_z_motion_mean = 0.0
        self._angular_velocity_z_motion_variance = 4.0
        
        self._kalman_odom = Odometry()

        self.get_logger().info(f"{self.get_name()} node initialized!")

    def _state_prediction(self):
        self._angular_velocity_z_mean = self._angular_velocity_z_mean + self._angular_velocity_z_motion_mean
        self._angular_velocity_z_variance = self._angular_velocity_z_variance + self._angular_velocity_z_motion_variance

    def _measurement_update(self):
        self._angular_velocity_z_mean = (self._imu_angular_velocity_z_variance * self._angular_velocity_z_mean + self._angular_velocity_z_variance * self._imu_angular_velocity_z_mean) / (self._angular_velocity_z_variance + self._imu_angular_velocity_z_variance)
        self._angular_velocity_z_variance = (self._angular_velocity_z_variance * self._imu_angular_velocity_z_variance) / (self._angular_velocity_z_variance + self._imu_angular_velocity_z_variance)

    def _odom_callback(self, odom: Odometry):
        self._kalman_odom = odom

        if self._is_first_ododm:
            self._odom_angular_velocity_z_last = odom.twist.twist.angular.z
            self._angular_velocity_z_mean = odom.twist.twist.angular.z
            self._is_first_ododm = False
            return
        
        self._angular_velocity_z_motion_mean = odom.twist.twist.angular.z - self._odom_angular_velocity_z_last
        if self._angular_velocity_z_motion_mean == math.inf or self._angular_velocity_z_motion_mean == math.nan:
            return 
        
        self._state_prediction()

        self._measurement_update()

        self._kalman_odom.twist.twist.angular.z = self._angular_velocity_z_mean
        self._odom_pub.publish(msg=self._kalman_odom)

    def _imu_callback(self, imu: Imu):
        self._imu_angular_velocity_z_mean = imu.angular_velocity.z

def main(args=None):
    try:
        rclpy.init(args=args)
        node = KalmanFilter()
        rclpy.spin(node=node)
        rclpy.shutdown()
    except KeyboardInterrupt:
        print("KeyboardInterrupt")
    finally:
        node.destroy_node()

if __name__ == "__main__":
    main()


